mark.c

```diff
@@ -829,6 +833,51 @@ GC_INNER mse * GC_mark_from(mse *mark_stack_top, mse *mark_stack,
           }
           if ((word)current_p > (word)limit) goto next_object;
         }
+#       else   /* !defined(__CHERI_PURE_CAPABILITY__) */
+        /* Try to prefetch the next pointer to be examined ASAP.        */
+        /* Check each pointer for validity before dereferencing         */
+        /* to prevent capability exceptions.                            */
+        /* Empirically, this also seems to help slightly without        */
+        /* prefetches, at least on linux/X86.  Presumably this loop     */
+        /* ends up with less register pressure, and gcc thus ends up    */
+        /* generating slightly better code.  Overall gcc code quality   */
+        /* for this loop is still not great.                            */
```

Comment:
I'd be tempted to skip this optimisation, making the loop simple. It probably doesn't apply to any of the CHERI targets anyway.

Reply:
I have for the moment removed the loop unroll and the PREFETCH from the code as the CHERI version is under a separate CHERI switch.